grails-app/controllers/au/org/ala/ecodata/CommentController.groovy

```diff
@@ -0,0 +1,161 @@
+package au.org.ala.ecodata
+
+import grails.converters.JSON
+import groovy.json.JsonSlurper
+
+class CommentController {
+    def outputService
+
+    def ignores = ["action","controller"]
+
+    def list() {
+        List comments = []
+        def sort = params.sort ? params.sort : "dateCreated"
+        def order = params.order ? params.order :  "desc"
+        def offset = params.start ? params.start : 0
+        def max = params.pageSize ? params.pageSize : 10
+        Boolean sortOrder = order == 'asc'? true : false;
+        String  entityId = params.entityId
+        String entityType = params.entityType
+        Integer total = 0;
+
+        if( !entityId || !entityType){
+            response.sendError(400, 'Insufficient parameters provided. Missing either entityId or entityType')
+        } else {
+//           todo: is there a better way to do count. is the below method potentially slow?
+            total = Comment.findAllWhere([ 'entityId':entityId, 'entityType':entityType, parent: null]).size();
```

Comment:
There are a couple of option. The best approach would be to push the count down to the db, rather than fetching the records then counting.

The easiest approach would be to use the GORM countBy... dynamic method: something like ```Comment.countByEntityIdAndEntityTypeAndParentIsNull(entityId, entityType)```. List the conditions in order of how restrictive they are (more restrictive first) to get the best performance.

Alternatively, you could hook into the mongo api and either use mongo's count() function or its aggregates. I doubt the performance difference (if there is one) would warrant that though.